src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h

```diff
@@ -0,0 +1,579 @@
+  static inline void generate_particle_lattice(
+      MPI_Comm comm
+    , ReadBoundsSelectionMode bounds_mode
+    , const Domain& domain
+    , GridT& grid
+    , const ParticleTypeMap& particle_type_map
+    , const ParticleRegions* particle_regions
+    , ParticleRegionCSG* region
+    , const GridCellValues* grid_cell_values
+    , const std::string* grid_cell_mask_name
+    , const double* grid_cell_mask_value
+    , const ScalarSourceTermInstance user_function
+    , double user_threshold
+    , const std::string& structure
+    , const std::vector<std::string>& types
+    , double sigma_noise
+    , const Vec3d& size
+    , double noise_cutoff
+    , const Vec3d& position_shift
+    , const std::string& void_mode
+    , const Vec3d& void_center
+    , double void_radius
+    , double void_porosity
+    , double void_mean_diameter
+    , ParticleTypeField )
+  {
+    using has_field_type_t = typename GridT:: template HasField < ParticleTypeField >;
+    static constexpr bool has_field_type = has_field_type_t::value;
+
+    using has_field_id_t = typename GridT:: template HasField < field::_id >;
+    static constexpr bool has_field_id = has_field_id_t::value;
+
+    using ParticleTupleIO = std::conditional_t< has_field_type
+                                              , onika::soatl::FieldTuple<field::_rx, field::_ry, field::_rz, field::_id, ParticleTypeField>
+                                              , onika::soatl::FieldTuple<field::_rx, field::_ry, field::_rz, field::_id>
+                                              >;
+
+    Vec3d lattice_size = size;
+    double noise_upper_bound = std::numeric_limits<double>::max();
+    if( noise_cutoff >= 0.0 ) noise_upper_bound = noise_cutoff;
+    else noise_upper_bound = std::min(lattice_size.x,std::min(lattice_size.y,lattice_size.z)) * 0.5;
+
+    // MPI Initialization
+    int rank=0, np=1;
+    MPI_Comm_rank(comm, &rank);
+    MPI_Comm_size(comm, &np);
+
+    // uint64_t n_particles = 0;
+
+    // Verification du nombre de types d'atomes en fonction du type de maille cubique voulu
+    // ***choix**  **nb_types**
+    //         SC   1  
+    //        BCC   2
+    //        FCC   4
+    //        HCP   4
+    //    DIAMOND   8
+    //   ROCKSALT   8
+    //   FLUORITE  12
+    //        C15  24
+    // PEROVSKITE   5
+    //         ST   1
+    //        BCT   2
+    //        FCT   4
+    //       2BCT   4
+
+    lout << std::defaultfloat
+         << "======= Lattice generator ======="<< std::endl
+         << "structure         = " << structure << std::endl
+         << "types             =";
+    for(const auto& s:types) lout <<" "<<s;
+    lout << std::endl
+         << "lattice cell size = "<< lattice_size << std::endl
+         << "position shift    = "<< position_shift <<std::endl
+         << "noise sigma       = "<< sigma_noise <<std::endl
+         << "noise cutoff      = "<< noise_cutoff <<std::endl;
+    
+    uint64_t n_particles_cell = 0.;
```

Comment:
Why not add a new LatticeCollection class  to improve readability and give every application the possibility of creating an atypical lattice?

Reply:
Yes, we can add a new LatticeCollection class I agree. I'll open an issue for that so we can move forward with this PR.

Reply:
aggreed, but this is let to a later dev